feat(topology/uniform_space): Abstract completions

PatrickMassot opened the pull request:
Define abstract completions, study their properties and derived constructions. Use this study in the concrete completion files, and for comparing completions constructed at different level of generality, especially for real numbers.

See the headers of `completion_pkg.lean` and `compare_reals.lean` for more details.

This comes from the perfectoid spaces project.

Commit:
feat(topology/uniform_space): Abstract completions

Define abstract completions, study their properties and derived
constructions. Use this study in the concrete completion files,
and for comparing completions constructed at different level of
generality, especially for real numbers.

This comes from the perfectoid spaces project.

jcommelin commented on src/topology/uniform_space/absolute_value.lean:
```diff
@@ -0,0 +1,73 @@
+/-!
+# Uniform structure induced by an absolute value
+
+We build an uniform space structure on a commutative ring `R` equipped with an absolute value into
```
```suggestion
We build a uniform space structure on a commutative ring `R` equipped with an absolute value into
```

jcommelin commented on src/topology/uniform_space/absolute_value.lean:
```diff
@@ -0,0 +1,73 @@
+`𝕜 = ℝ`, we get the same thing as the metric space construction, and the general construction
+follows exactly the same path.
+
+Note that we import `data.real.cau_seq` because this is where absolute values are defined, but
```
```suggestion
## Implementation details

Note that we import `data.real.cau_seq` because this is where absolute values are defined, but
```
Maybe add this header?

jcommelin commented on src/topology/uniform_space/compare_reals.lean:
```diff
@@ -0,0 +1,114 @@
+# Comparison of Cauchy reals and Bourbaki reals
+
+In `data.real.basic` real numbers are defined using the so called Cauchy construction (although
+it is due to Georg Canter). More precisely, this construction applies to commutative rings equipped
```
```suggestion
it is due to Georg Cantor). More precisely, this construction applies to commutative rings equipped
```

jcommelin commented on src/topology/uniform_space/completion_pkg.lean:
```diff
@@ -0,0 +1,313 @@
+/-!
+# Abstract theory of Hausdorff completions of uniform spaces
+
+This file caracterizes Hausdorff completions of a uniform space α as complete Hausdorff spaces
```
```suggestion
This file characterizes Hausdorff completions of a uniform space α as complete Hausdorff spaces
```

jcommelin commented on src/topology/uniform_space/completion_pkg.lean:
```diff
@@ -0,0 +1,313 @@
+
+## References
+
+We don't know any traditionnal text discussing this. Real world mathematics simply silently
```
```suggestion
We don't know any traditional text discussing this. Real world mathematics simply silently
```

jcommelin reviewed (commented):
Some typos in the docstrings

PatrickMassot committed:
Apply suggestions from code review by Johan

Co-Authored-By: Johan Commelin <johan@commelin.net>

PatrickMassot committed:
Fix build.

When I created the PR, github complained it couldn't automatically
merge, and I was not careful enough when I merged...

Commit:
chore(topology/uniform_space): rename completion_pkg

sgouezel commented on src/topology/uniform_space/compare_reals.lean:
```diff
@@ -0,0 +1,114 @@
+def Bourbaki_pkg : abstract_completion Q := completion.cpkg
+
+/-- The equivalence between Bourbaki and Cauchy reals-/
+noncomputable def compare_equiv : Bourbakiℝ ≃ ℝ :=
```
This statement and the next ones are in the root namespace, as is the definition of `Q` above. Since these are quite general, I would put most of this file in a dedicated namespace to avoid clashes.

PatrickMassot replied:
This file is very unlikely to be imported anywhere, but I'll create a namespace anyway.

Commit:
fix(compare_reals): create namespace

Commit:
Fix build

mergify[bot] committed:
Merge branch 'master' into completion_pkg
